File: javascript/leetcode/859. Buddy Strings.js
/**
 * @param {string} A
 * @param {string} B
 * @return {boolean}
 */
var buddyStrings = function (A, B) {
  if (A.length != B.length) return false;

  if (A === B) {
    const seen = new Set();

    for (let i of A) seen.add(i);

    if (seen.size < A.length) return true;
    return false;
  }

  const diff = [];
  for (let i = 0; i < A.length; ++i) {
    if (A[i] !== B[i]) {
      diff.push(i);
      if (diff.length > 2) break;
    }
  }

  if (
    diff.length === 2 &&
    A[diff[0]] === B[diff[1]] &&
    B[diff[0]] === A[diff[1]]
  )
    return true;

  return false;
};

File: c++/leetcode/859. Buddy Strings.cpp
class Solution
{
public:
  bool buddyStrings(string A, string B)
  {
    if (A.size() != B.size())
      return false;

    if (A == B)
    {
      unordered_set<char> seen;
      for (auto i : A)
        seen.insert(i);

      if (seen.size() < A.size())
        return true;
      return false;
    }

    vector<int> diff;
    for (int i = 0; i < A.size(); ++i)
    {
      if (A[i] != B[i])
      {
        diff.push_back(i);
        if (diff.size() > 2)
          break;
      }
    }

    if (diff.size() == 2 && A[diff[0]] == B[diff[1]] &&
        B[diff[0]] == A[diff[1]])
      return true;

    return false;
  }
};